File: mopub_flutter-1.2.2+1/android/src/main/java/com/mopub_flutter/MopubRewardedVideoAdPlugin.java
package com.mopub_flutter;

import androidx.annotation.NonNull;

import com.mopub.common.MoPubReward;
import com.mopub.mobileads.MoPubErrorCode;
import com.mopub.mobileads.MoPubRewardedVideoListener;
import com.mopub.mobileads.MoPubRewardedVideos;

import java.util.HashMap;
import java.util.Set;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.PluginRegistry;

class MopubRewardedVideoAdPlugin implements MethodChannel.MethodCallHandler {

    private PluginRegistry.Registrar registrar;

    MopubRewardedVideoAdPlugin(PluginRegistry.Registrar registrar) {
        this.registrar = registrar;
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        final HashMap args = (HashMap) call.arguments;
        final String adUnitId = (String) args.get("adUnitId");

        switch (call.method) {
            case MopubConstants.SHOW_REWARDED_VIDEO_METHOD:

                if (MoPubRewardedVideos.hasRewardedVideo(adUnitId))
                    MoPubRewardedVideos.showRewardedVideo(adUnitId);

                result.success(null);
                break;
            case MopubConstants.LOAD_REWARDED_VIDEO_METHOD:

                final MethodChannel adChannel = new MethodChannel(registrar.messenger(),
                        MopubConstants.REWARDED_VIDEO_CHANNEL + "_" + adUnitId);
                MoPubRewardedVideos.setRewardedVideoListener(new RewardedVideoListener(adChannel));
                if (MoPubRewardedVideos.hasRewardedVideo(adUnitId))
                    return;
                MoPubRewardedVideos.loadRewardedVideo(adUnitId);

                result.success(null);
                break;
            case MopubConstants.HAS_REWARDED_VIDEO_METHOD:
                boolean hasAd = MoPubRewardedVideos.hasRewardedVideo(adUnitId);
                result.success(hasAd);
                break;
            default:
                result.notImplemented();
        }
    }

    class RewardedVideoListener implements MoPubRewardedVideoListener {

        MethodChannel channel;
        boolean didReceiveReward = false;
        int rewardAmount = 0;

        RewardedVideoListener(MethodChannel channel) {
            this.channel = channel;
        }

        @Override
        public void onRewardedVideoLoadSuccess(@NonNull String adUnitId) {
            HashMap<String, Object> args = new HashMap<>();
            args.put("adUnitId", adUnitId);

            channel.invokeMethod(MopubConstants.LOADED_METHOD, args);
        }

        @Override
        public void onRewardedVideoLoadFailure(@NonNull String adUnitId, @NonNull MoPubErrorCode errorCode) {
            HashMap<String, Object> args = new HashMap<>();
            args.put("adUnitId", adUnitId);
            args.put("errorCode", errorCode.toString());

            channel.invokeMethod(MopubConstants.ERROR_METHOD, args);
        }

        @Override
        public void onRewardedVideoStarted(@NonNull String adUnitId) {
            HashMap<String, Object> args = new HashMap<>();
            args.put("adUnitId", adUnitId);

            channel.invokeMethod(MopubConstants.DISPLAYED_METHOD, args);
        }

        @Override
        public void onRewardedVideoPlaybackError(@NonNull String adUnitId, @NonNull MoPubErrorCode errorCode) {
            HashMap<String, Object> args = new HashMap<>();
            args.put("adUnitId", adUnitId);
            args.put("errorCode", errorCode.toString());

            channel.invokeMethod(MopubConstants.REWARDED_PLAYBACK_ERROR, args);
        }

        @Override
        public void onRewardedVideoClicked(@NonNull String adUnitId) {
            HashMap<String, Object> args = new HashMap<>();
            args.put("adUnitId", adUnitId);

            channel.invokeMethod(MopubConstants.CLICKED_METHOD, args);
        }

        @Override
        public void onRewardedVideoClosed(@NonNull String adUnitId) {
            HashMap<String, Object> args = new HashMap<>();
            args.put("adUnitId", adUnitId);

            channel.invokeMethod(MopubConstants.DISMISSED_METHOD, args);
            if (didReceiveReward) {
                didReceiveReward = false;
                final HashMap<String, Object> rewardArgs = new HashMap<>();
                rewardArgs.put("reward", rewardAmount);
                channel.invokeMethod(MopubConstants.GRANT_REWARD, rewardArgs);
            }

        }

        @Override
        public void onRewardedVideoCompleted(@NonNull Set<String> adUnitIds, @NonNull MoPubReward reward) {
            didReceiveReward = true;
            rewardAmount = reward.getAmount();
        }
    }
}

File: mopub_flutter-1.2.2+1/ios/Classes/MopubRewardedVideoAdPlugin.swift
//
//  MopubRewardedVideoAdPlugin.swift
//  mopub_flutter
//
//  Created by Rohit Kulkarni on 1/7/20.
//

import Foundation
import MoPub

public class MopubRewardedVideoAdPlugin: NSObject, FlutterPlugin {
    
    fileprivate var pluginRegistrar: FlutterPluginRegistrar?
//    Need to keep strong reference to delegate because MoPub by default keeps weak reference
    fileprivate var delegate: MPRewardedVideoDelegate?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = MopubRewardedVideoAdPlugin()
        instance.pluginRegistrar = registrar
        let channel = FlutterMethodChannel(name: MopubConstants.REWARDED_VIDEO_CHANNEL, binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String : Any] ?? [:]
        let adUnitId = args["adUnitId"] as? String ?? "ad_unit_id"
        
        switch call.method {
        case MopubConstants.SHOW_REWARDED_VIDEO_METHOD:
            if(MPRewardedVideo.hasAdAvailable(forAdUnitID: adUnitId)) {
                let reward = MPRewardedVideo.selectedReward(forAdUnitID: adUnitId)
                let rootViewController = UIApplication.shared.keyWindow?.rootViewController
                MPRewardedVideo.presentAd(forAdUnitID: adUnitId, from: rootViewController, with: reward)
            }
            result(nil)
            break;
        case MopubConstants.LOAD_REWARDED_VIDEO_METHOD:            
            let channelName = "\(MopubConstants.REWARDED_VIDEO_CHANNEL)_\(adUnitId)"
            let adChannel = FlutterMethodChannel(name: channelName, binaryMessenger: pluginRegistrar!.messenger())
            self.delegate = MopubRewardedVideoDelegate(channel: adChannel)
            MPRewardedVideo.removeDelegate(forAdUnitId: adUnitId)
            MPRewardedVideo.setDelegate(delegate, forAdUnitId: adUnitId)
            MPRewardedVideo.loadAd(withAdUnitID: adUnitId, withMediationSettings: nil)
            
            result(nil)
            break
        case MopubConstants.HAS_REWARDED_VIDEO_METHOD:
            let hasAd = MPRewardedVideo.hasAdAvailable(forAdUnitID: adUnitId)
            result(hasAd);
            break;
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

class MopubRewardedVideoDelegate: NSObject, MPRewardedVideoDelegate {
    let channel: FlutterMethodChannel
    var didReceiveReward: Bool
    var rewardAmount: NSNumber
    
    init(channel: FlutterMethodChannel) {
        self.channel = channel
        self.didReceiveReward = false
        self.rewardAmount = 0
    }
    
    func rewardedVideoAdDidLoad(forAdUnitID adUnitID: String!) {
        var args:[String:Any] = [:]
        args["adUnitId"] = adUnitID
        channel.invokeMethod(MopubConstants.LOADED_METHOD, arguments: args)
    }
    
    func rewardedVideoAdDidFailToLoad(forAdUnitID adUnitID: String!, error: Error!) {
        var args:[String:Any] = [:]
        args["adUnitId"] = adUnitID
        args["errorCode"] = error.localizedDescription
        channel.invokeMethod(MopubConstants.ERROR_METHOD, arguments: args)
    }
    
    func rewardedVideoAdWillAppear(forAdUnitID adUnitID: String!) {    
        var args:[String:Any] = [:]
        args["adUnitId"] = adUnitID
        channel.invokeMethod(MopubConstants.DISPLAYED_METHOD, arguments: args)
    }
    
    func rewardedVideoAdDidFailToPlay(forAdUnitID adUnitID: String!, error: Error!) {
        var args:[String:Any] = [:]
        args["adUnitId"] = adUnitID
        args["errorCode"] = error.localizedDescription
        channel.invokeMethod(MopubConstants.REWARDED_PLAYBACK_ERROR, arguments: args)
    }
    
    func rewardedVideoAdDidDisappear(forAdUnitID adUnitID: String!) {
        var args:[String:Any] = [:]
        args["adUnitId"] = adUnitID
        channel.invokeMethod(MopubConstants.DISMISSED_METHOD, arguments: args)
        if(didReceiveReward) {
            didReceiveReward = false
            var args:[String:Any] = [:]
            args["reward"] = rewardAmount
            channel.invokeMethod(MopubConstants.GRANT_REWARD, arguments: args)
        }
    }
    
    func rewardedVideoAdDidReceiveTapEvent(forAdUnitID adUnitID: String!) {
        var args:[String:Any] = [:]
        args["adUnitId"] = adUnitID
        channel.invokeMethod(MopubConstants.CLICKED_METHOD, arguments: args)
    }
    
    func rewardedVideoAdShouldReward(forAdUnitID adUnitID: String!, reward: MPRewardedVideoReward!) {
        self.didReceiveReward = true;
        self.rewardAmount = reward.amount
    }
    
    func rewardedVideoAdDidExpire(forAdUnitID adUnitID: String!) {
        MPRewardedVideo.loadAd(withAdUnitID: adUnitID, withMediationSettings: nil)
    }
    
    
    
    
}
